File: internal/capability/validate.go
package capability

import "github.com/aigc-pool/aigc-pool/internal/domain"

// Submission 是一次待校验的任务提交，字段与 POST /api/tasks 的请求体对应。
//
// 注意**没有 mode 字段**：传了 Inputs 就是图生图/图生视频，没传就是文生，
// 隐式分流由服务端按输入槽是否为空判定，前端不参与。
type Submission struct {
	ModelID string
	Prompt  string
	Inputs  map[string][]string
	Params  map[string]any
}

// Validator 对一次提交做服务端校验。
//
// 前端已经按 schema 渲染并预校验过一遍，但那只是体验优化——**服务端必须独立再校验一遍**，
// 因为请求可以绕过前端直接构造。这里是唯一可信的那一遍。
//
// 校验范围:
//   - 未知的 params key / inputs slot 一律拒绝（防止污染上游请求体）；
//   - 每个参数的取值落在其 ParamSpec 声明的域内（select 在 options 里、
//     stepper/slider 在 [Min,Max] 且对齐 Step、textarea 不超 MaxLength）；
//   - VisibleWhen 为假的参数**不允许出现**在 Params 里（前端约定不提交它们，
//     出现即代表来路可疑或前端有 bug）；
//   - 必填输入槽有值，可选槽的数量落在 [MinCount, MaxCount]，MIME 在 Accept 白名单内；
//   - prompt 长度不超过 LimitSpec.PromptMaxLength；
//   - ParamSpec 结构自洽（见 ValidateSchema）。
//
// Validate 失败时返回的 FieldError 列表直接作为 TaskError.FieldErrors 下发，
// Key 必须是 ParamSpec.Key 或 InputSlotSpec.Key——前端靠它高亮出错的那枚芯片。
// 因此实现应**收集全部错误再返回**，而不是遇到第一个就短路：一次提交里三个参数
// 都填错，用户应当一次看到三个红点，而不是改一个试一次。
//
// ValidateSchema 校验 ModelCapabilitySchema 自身是否自洽，用于管理后台
// POST/PATCH /api/admin/models 落库前把关。包注释里说过，ParamSpec 是所有控件
// 字段的并集结构，类型系统不阻止给 toggle 填 Options——那条约束就在这里兑现：
// 每种 Control 必须携带它要求的字段（select 要 Options、stepper 要 Min/Max/Step、
// compound 要 Fields + DisplayTemplate 且自身 Default 为 nil），
// 且不得携带其他控件专属的字段。写错配置应在保存时就被拒，
// 而不是等某个用户提交任务时才炸。
type Validator interface {
	Validate(schema ModelCapabilitySchema, sub Submission) []domain.FieldError
	ValidateSchema(schema ModelCapabilitySchema) []domain.FieldError
}

File: frontend/src/schema/validate.ts
import type { ModelCapabilitySchema, OptionItem, ParamSpec } from './types';
import type { FormValues, InputValues } from './evaluate';
import { isVisible, leafSpecs } from './form';
import { fallbackOptions, isKnownControl } from './controls';

export type FieldErrors = Record<string, string>;

function optionsOf(spec: ParamSpec): OptionItem[] | null {
  if (spec.control === 'select' || spec.control === 'aspect_select') return spec.options;
  return isKnownControl(spec) ? null : fallbackOptions(spec);
}

/** 单个参数值是否满足 schema 约束。切模型迁移与提交前校验共用这一份判定 */
export function isValidValue(spec: ParamSpec, value: unknown): boolean {
  const options = optionsOf(spec);
  if (options) return options.some((o) => o.value === value && !o.disabled);

  switch (spec.control) {
    case 'stepper':
    case 'slider': {
      const n = Number(value);
      return Number.isFinite(n) && n >= spec.min && n <= spec.max;
    }
    case 'seed':
      if (value === null) return true;
      return Number.isFinite(Number(value)) && Number(value) >= spec.min && Number(value) <= spec.max;
    case 'toggle':
      return typeof value === 'boolean';
    case 'textarea':
      return typeof value === 'string' && value.length <= spec.max_length;
    case 'compound':
      return true;
    default:
      // 未知 control 且无 options：只做透传，不判定
      return true;
  }
}

/**
 * 前端校验是体验不是安全边界（frontend-design.md §3.5）——
 * 后端仍会完整再校验一遍，这里只是不让用户盲提交。
 */
export function validateForm(
  model: ModelCapabilitySchema,
  prompt: string,
  values: FormValues,
  inputs: InputValues,
): FieldErrors {
  const errors: FieldErrors = {};

  if (prompt.length > model.limits.prompt_max_length) {
    errors.prompt = `提示词最长 ${model.limits.prompt_max_length} 字，当前 ${prompt.length}`;
  }

  for (const slot of model.inputs) {
    const files = inputs[slot.key] ?? [];
    if (slot.required && files.length < Math.max(1, slot.min_count)) {
      errors[slot.key] = `${slot.label}为必填`;
    } else if (files.length > slot.max_count) {
      errors[slot.key] = `${slot.label}最多 ${slot.max_count} 个`;
    } else if (slot.requires_slot && files.length > 0 && !(inputs[slot.requires_slot] ?? []).length) {
      const dep = model.inputs.find((s) => s.key === slot.requires_slot);
      errors[slot.key] = `${slot.label}需要同时提供${dep?.label ?? slot.requires_slot}`;
    }
  }

  for (const spec of leafSpecs(model.params)) {
    if (!isVisible(spec, values, inputs)) continue;
    if (!isValidValue(spec, values[spec.key])) {
      errors[spec.key] = `${spec.label}的取值不合法`;
    }
  }

  return errors;
}

/** 上传前的本地预校验，避免传完才报错 */
export async function validateFile(
  slot: ModelCapabilitySchema['inputs'][number],
  file: File,
): Promise<string | null> {
  if (slot.accept.length && !slot.accept.includes(file.type)) {
    return `仅支持 ${slot.accept.map((a) => a.replace(/^\w+\//, '')).join(' / ')}`;
  }
  if (file.size > slot.max_bytes) {
    return `文件不能超过 ${Math.round(slot.max_bytes / 1024 / 1024)}MB`;
  }
  if (slot.kind !== 'image' || (!slot.min_pixels && !slot.max_pixels)) return null;

  const size = await readImageSize(file);
  if (!size) return null;
  if (slot.min_pixels && (size.width < slot.min_pixels.width || size.height < slot.min_pixels.height)) {
    return `尺寸不能小于 ${slot.min_pixels.width}×${slot.min_pixels.height}`;
  }
  if (slot.max_pixels && (size.width > slot.max_pixels.width || size.height > slot.max_pixels.height)) {
    return `尺寸不能大于 ${slot.max_pixels.width}×${slot.max_pixels.height}`;
  }
  return null;
}

function readImageSize(file: File): Promise<{ width: number; height: number } | null> {
  return new Promise((resolve) => {
    const url = URL.createObjectURL(file);
    const img = new Image();
    img.onload = () => {
      URL.revokeObjectURL(url);
      resolve({ width: img.naturalWidth, height: img.naturalHeight });
    };
    img.onerror = () => {
      URL.revokeObjectURL(url);
      resolve(null);
    };
    img.src = url;
  });
}
